install.sh

```diff
@@ -780,10 +1171,87 @@ function register_sandcats_name() {
-    register_sandcats_name
+    sandcats_register_name
     return
   fi
 }
 
+sandcats_generate_keys() {
+    # The Sandcats service places its authentication files in $DIR/var/sandcats.
+    if [ -f var/sandcats/id_rsa.private_combined ] ; then
+        return
+    fi
+
+    # The openssl key generation process can take a few seconds, so we
+    # print a ... while that happens.
+    echo -n '...'
+
+    # We are already in $DIR. It's important to make it mode 0700
+    # because we store TLS client authentication keys here.
+    mkdir -p -m 0700 var/sandcats
+    chmod 0700 var/sandcats
+
+    # Generate key for client certificate. OpenSSL will read from
+    # /dev/urandom by default, so this won't block. We abuse the ``
+    # operator so we can have inline comments in a multi-line command.
+    openssl \
+        req `# Invoke OpenSSL's PKCS#10 X.509 bits.` \
+        -new `# Create a new certificate/request.` \
+        -newkey rsa:4096 `# Create a new RSA key of length 4096 bits.` \
+        -days 3650 `# Make the self-signed cert valid for 10 years.` \
+        -nodes `# no DES -- that is, do not encrypt the key at rest.` \
+        -x509 `# Output a certificate, rather than a signing request.` \
+        `# Sandcats ignores the subject in the certificate; use` \
+        `# OpenSSL defaults.` \
+        -subj "/C=AU/ST=Some-State/O=Internet Widgits Pty Ltd" \
+        -keyout var/sandcats/id_rsa `# Store the resulting RSA private key in id_rsa` \
+        -out var/sandcats/id_rsa.pub `# Store the resulting certificate in id_rsa.pub` \
+        2>/dev/null `# Silence the progress output.`
+
+    # We combine these two things into one glorious all-inclusive file
+    # for the `curl` command. It is just as private as id_rsa.
+    cat var/sandcats/id_rsa var/sandcats/id_rsa.pub > var/sandcats/id_rsa.private_combined
+
+    # Set filesystem permissions, in case the files get copied
+    # into the wrong place later.
+    chmod 0600 var/sandcats/id_rsa var/sandcats/id_rsa.pub var/sandcats/id_rsa.private_combined
+
+    # Go to the start of the line, before the "..." that we
+    # left on the screen, allowing future echo statements to
+    # overwrite it.
+    echo -ne '\r'
+}
+
+# Now that the steps exist as functions, run them in an order that
+# would result in a working install.
+detect_current_uid
+handle_args "$@"
+assert_on_terminal
+assert_linux_x86_64
+assert_usable_kernel
+assert_userns_clone
+assert_dependencies
+assert_valid_bundle_file
+detect_init_system
+choose_install_mode
+choose_external_or_internal
+choose_install_dir
+load_existing_settings
+choose_server_user_if_needed
+create_server_user_if_needed
+configure_auto_updates
+configure_dev_accounts
+configure_hostnames
+save_config
+download_latest_bundle_and_extract_if_needed
+extract_bundle_if_provided
+make_runtime_directories
+set_permissions
+install_sandstorm_symlinks
+ask_about_starting_at_boot
+configure_start_at_boot_if_desired
+print_success
+}
```

Comment:
Right-o. This close-brace matches the `_` function definition at the top of the file.